File: Sources/NBT/NBTTypes.swift
import Foundation

enum NBTEncoding: Equatable {
    case bigEndian
    case littleEndian
    case littleEndianVarInt
}

enum NBTTagType: UInt8, CaseIterable {
    case end = 0
    case byte = 1
    case short = 2
    case int = 3
    case long = 4
    case float = 5
    case double = 6
    case byteArray = 7
    case string = 8
    case list = 9
    case compound = 10
    case intArray = 11
    case longArray = 12
}

struct NBTNamedTag {
    var name: String
    var value: NBTValue
}

indirect enum NBTValue {
    case byte(Int8)
    case short(Int16)
    case int(Int32)
    case long(Int64)
    case float(Float)
    case double(Double)
    case byteArray(Data)
    case string(String)
    case list(NBTTagType, [NBTValue])
    case compound([NBTNamedTag])
    case intArray([Int32])
    case longArray([Int64])

    var type: NBTTagType {
        switch self {
        case .byte: return .byte
        case .short: return .short
        case .int: return .int
        case .long: return .long
        case .float: return .float
        case .double: return .double
        case .byteArray: return .byteArray
        case .string: return .string
        case .list: return .list
        case .compound: return .compound
        case .intArray: return .intArray
        case .longArray: return .longArray
        }
    }

    var summary: String {
        switch self {
        case .byte(let value): return String(value)
        case .short(let value): return String(value)
        case .int(let value): return String(value)
        case .long(let value): return String(value)
        case .float(let value): return String(value)
        case .double(let value): return String(value)
        case .byteArray(let value): return "ByteArray[\(value.count)]"
        case .string(let value): return value
        case .list(_, let values): return "List[\(values.count)]"
        case .compound(let values): return "Compound{\(values.count)}"
        case .intArray(let values): return "IntArray[\(values.count)]"
        case .longArray(let values): return "LongArray[\(values.count)]"
        }
    }

    var children: [NBTNamedTag] {
        switch self {
        case .compound(let tags): return tags
        case .list(_, let values): return values.enumerated().map { NBTNamedTag(name: "[\($0.offset)]", value: $0.element) }
        default: return []
        }
    }

    func compoundValue(named name: String) -> NBTValue? {
        guard case .compound(let tags) = self else { return nil }
        return tags.first(where: { $0.name == name })?.value
    }

    func stringValue(named name: String) -> String? {
        guard let value = compoundValue(named: name) else { return nil }
        if case .string(let text) = value { return text }
        return nil
    }

    func intValue(named name: String) -> Int32? {
        guard let value = compoundValue(named: name) else { return nil }
        switch value {
        case .int(let number): return number
        case .short(let number): return Int32(number)
        case .byte(let number): return Int32(number)
        default: return nil
        }
    }
}

struct NBTDocument {
    var rootName: String
    var root: NBTValue
}

extension NBTTagType {
    var displayName: String {
        switch self {
        case .end: return "End"
        case .byte: return "Byte"
        case .short: return "Short"
        case .int: return "Int"
        case .long: return "Long"
        case .float: return "Float"
        case .double: return "Double"
        case .byteArray: return "ByteArray"
        case .string: return "String"
        case .list: return "List"
        case .compound: return "Compound"
        case .intArray: return "IntArray"
        case .longArray: return "LongArray"
        }
    }
}

File: Windows/MCBEEditor.Core/Nbt/NbtTypes.cs
using System.Globalization;
using System.Text;

namespace MCBEEditor.Core.Nbt;

public enum NbtEncoding
{
    BigEndian,
    LittleEndian,
    LittleEndianVarInt
}

public enum NbtTagType : byte
{
    End = 0,
    Byte = 1,
    Short = 2,
    Int = 3,
    Long = 4,
    Float = 5,
    Double = 6,
    ByteArray = 7,
    String = 8,
    List = 9,
    Compound = 10,
    IntArray = 11,
    LongArray = 12
}

public sealed record NbtNamedTag(string Name, NbtValue Value);

public abstract record NbtValue
{
    public abstract NbtTagType Type { get; }

    public virtual string Summary => Type.ToString();

    public virtual IReadOnlyList<NbtNamedTag> Children => Array.Empty<NbtNamedTag>();

    public NbtValue? CompoundValue(string name)
    {
        if (this is not NbtCompoundValue compound)
            return null;

        return compound.Tags.FirstOrDefault(t => t.Name == name)?.Value;
    }

    public NbtValue? CompoundValueIgnoreCase(params string[] names)
    {
        if (this is not NbtCompoundValue compound || names.Length == 0)
            return null;

        foreach (var name in names)
        {
            var exact = compound.Tags.FirstOrDefault(t => t.Name == name);
            if (exact is not null)
                return exact.Value;
        }

        var set = new HashSet<string>(names, StringComparer.OrdinalIgnoreCase);
        return compound.Tags.FirstOrDefault(t => set.Contains(t.Name))?.Value;
    }

    public string? StringValue(string name) => CompoundValue(name) is NbtStringValue value ? value.Value : null;

    public int? IntValue(string name) => CompoundValue(name) switch
    {
        NbtByteValue value => value.Value,
        NbtShortValue value => value.Value,
        NbtIntValue value => value.Value,
        _ => null
    };

    public long? IntegerValue() => this switch
    {
        NbtByteValue value => value.Value,
        NbtShortValue value => value.Value,
        NbtIntValue value => value.Value,
        NbtLongValue value => value.Value,
        NbtFloatValue value => FloatingIntegerValue(value.Value),
        NbtDoubleValue value => FloatingIntegerValue(value.Value),
        _ => null
    };

    private static long? FloatingIntegerValue(double value)
    {
        if (!double.IsFinite(value)) return null;
        var truncated = Math.Truncate(value);
        // 2^63 is the first Double outside Int64's positive range. Using
        // long.MaxValue directly in a Double comparison is unsafe because it
        // rounds to 2^63 as well.
        if (truncated < -9_223_372_036_854_775_808.0 || truncated >= 9_223_372_036_854_775_808.0)
            return null;
        return (long)truncated;
    }
}

public sealed record NbtByteValue(sbyte Value) : NbtValue
{
    public override NbtTagType Type => NbtTagType.Byte;
    public override string Summary => Value.ToString(CultureInfo.InvariantCulture);
}

public sealed record NbtShortValue(short Value) : NbtValue
{
    public override NbtTagType Type => NbtTagType.Short;
    public override string Summary => Value.ToString(CultureInfo.InvariantCulture);
}

public sealed record NbtIntValue(int Value) : NbtValue
{
    public override NbtTagType Type => NbtTagType.Int;
    public override string Summary => Value.ToString(CultureInfo.InvariantCulture);
}

public sealed record NbtLongValue(long Value) : NbtValue
{
    public override NbtTagType Type => NbtTagType.Long;
    public override string Summary => Value.ToString(CultureInfo.InvariantCulture);
}

public sealed record NbtFloatValue(float Value) : NbtValue
{
    public override NbtTagType Type => NbtTagType.Float;
    public override string Summary => Value.ToString("R", CultureInfo.InvariantCulture);
}

public sealed record NbtDoubleValue(double Value) : NbtValue
{
    public override NbtTagType Type => NbtTagType.Double;
    public override string Summary => Value.ToString("R", CultureInfo.InvariantCulture);
}

public sealed record NbtByteArrayValue(byte[] Value) : NbtValue
{
    public override NbtTagType Type => NbtTagType.ByteArray;
    public override string Summary => $"ByteArray[{Value.Length}]";
}

public sealed record NbtStringValue(string Value) : NbtValue
{
    public override NbtTagType Type => NbtTagType.String;
    public override string Summary => NbtRawStringCodec.DisplayText(Value);
}

public sealed record NbtListValue(NbtTagType ElementType, IReadOnlyList<NbtValue> Values) : NbtValue
{
    public override NbtTagType Type => NbtTagType.List;
    public override string Summary => $"List[{Values.Count}]";
    public override IReadOnlyList<NbtNamedTag> Children => Values.Select((value, index) => new NbtNamedTag($"[{index}]", value)).ToArray();
}

public sealed record NbtCompoundValue(IReadOnlyList<NbtNamedTag> Tags) : NbtValue
{
    public override NbtTagType Type => NbtTagType.Compound;
    public override string Summary => $"Compound{{{Tags.Count}}}";
    public override IReadOnlyList<NbtNamedTag> Children => Tags;
}

public sealed record NbtIntArrayValue(IReadOnlyList<int> Values) : NbtValue
{
    public override NbtTagType Type => NbtTagType.IntArray;
    public override string Summary => $"IntArray[{Values.Count}]";
}

public sealed record NbtLongArrayValue(IReadOnlyList<long> Values) : NbtValue
{
    public override NbtTagType Type => NbtTagType.LongArray;
    public override string Summary => $"LongArray[{Values.Count}]";
}

public sealed record NbtDocument(string RootName, NbtValue Root);

public static class NbtRawStringCodec
{
    private const string Sentinel = "\U000F0000\U000F0001MCBE_RAW_NBT_STRING:";
    private static readonly UTF8Encoding StrictUtf8 = new(false, true);

    public static string Decode(ReadOnlySpan<byte> bytes)
    {
        try
        {
            return StrictUtf8.GetString(bytes);
        }
        catch (DecoderFallbackException)
        {
            return Sentinel + Convert.ToHexString(bytes).ToLowerInvariant();
        }
    }

    public static byte[] Encode(string value)
    {
        if (TryRawBytes(value, out var raw))
            return raw;
        return Encoding.UTF8.GetBytes(value);
    }

    public static bool TryRawBytes(string value, out byte[] bytes)
    {
        bytes = Array.Empty<byte>();
        if (!value.StartsWith(Sentinel, StringComparison.Ordinal))
            return false;

        var hex = value[Sentinel.Length..];
        if ((hex.Length & 1) != 0)
            return false;

        try
        {
            bytes = Convert.FromHexString(hex);
            // The sentinel is reserved for byte sequences that are not UTF-8.
            _ = StrictUtf8.GetString(bytes);
            bytes = Array.Empty<byte>();
            return false;
        }
        catch (DecoderFallbackException)
        {
            return true;
        }
        catch (FormatException)
        {
            bytes = Array.Empty<byte>();
            return false;
        }
    }

    public static string DisplayText(string value)
    {
        if (!TryRawBytes(value, out var bytes))
            return value;

        return $"RawString[{bytes.Length}] {string.Join(' ', bytes.Select(b => b.ToString("x2", CultureInfo.InvariantCulture)))}";
    }
}
